### PR TITLE
ICMSLST-726 Add tests for FA-DFL applicant views.

### PR DESCRIPTION
**Pull request by MattHolmes123**
Changes:
  - Add django command to add dummy data to test database
  - Add top level fixtures for common things
  - Added dummy file upload handler
  - Add tests for the following endpoints:
    - edit_dfl
    - add_goods_certificate
    - edit_goods_certificate
    - submit_dfl

### CODE REVIEW

File: web/utils/validation.py [hunk 1]
```diff
@@ -42,6 +42,13 @@ def has_errors(self) -> bool:
     def add(self, page_errors: PageErrors) -> None:
         self.page_errors.append(page_errors)
 
+    def get_page_errors(self, page_name) -> Optional[PageErrors]:
```
Comment (oskusalerma) [inline web/utils/validation.py:45]: `page_name: str`

File: web/tests/helpers.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from web.utils.validation import ApplicationErrors
+
+
+def has_page_errors(errors: ApplicationErrors, page_name: str, error_field_names: list[str]):
```
Comment (oskusalerma) [inline web/tests/helpers.py:4]: check_page_errors

File: web/tests/file_upload_handler.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+class DummyFileUploadHandler(TemporaryFileUploadHandler):
+    def receive_data_chunk(self, raw_data, start):
+        return super().receive_data_chunk(raw_data, start)
+
+    def file_complete(self, file_size):
+        file = super().file_complete(file_size)
+
+        # Set extra attributes the real file handlers add.
+        file.original_name = f"original_name: {self.file_name}"
+        file.file_size = 1024
```
Comment (oskusalerma) [inline web/tests/file_upload_handler.py:13]: use the passed-in actual file size

File: web/tests/domains/case/_import/fa/conftest.py [hunk 4]
```diff
@@ -0,0 +1 @@
+# TODO: Add tests for common fa code (views)
```
Comment (oskusalerma) [inline web/tests/domains/case/_import/fa/conftest.py:1]: please delete empty/unneeded files

File: run-tests.sh [hunk 5]
```diff
@@ -13,3 +13,12 @@ export ICMS_DEBUG=False
 # --create-db" to force re-creation of the test database.
 
 docker-compose run --rm web pytest --tb=short --dist=loadfile --tx=4*popen "$@"
+
+# With coverage
+# docker-compose run --rm web pytest --tb=short --cov=web/domains/case/_import/fa_dfl --cov-report term-missing:skip-covered --dist=loadfile --tx=4*popen "$@"
+
+# Single runner for speed (When only running a subset of the tests):
```
Comment (oskusalerma) [inline run-tests.sh:20]: no need for single runners, add a comment saying "run with --dist=no when testing a single file"

File: web/tests/conftest.py [hunk 6]
```diff
@@ -17,3 +22,57 @@ def instrumented_render(template_object, *args, **kwargs):
...
+    """
+
+    with django_db_blocker.unblock():
+        call_command("add_test_data")
+
+
+# Read this to give all tests db access by default
+# https://pytest-django.readthedocs.io/en/latest/faq.html#how-can-i-give-database-access-to-all-my-tests-without-the-django-db-marker
+@pytest.fixture(autouse=True)
+def enable_db_access_for_all_fixtures(db):
```
Comment (oskusalerma) [inline web/tests/conftest.py:42]: test that lower-down tests don't require db (some of them don't)

File: web/tests/conftest.py [hunk 7]
```diff
@@ -17,3 +22,57 @@ def instrumented_render(template_object, *args, **kwargs):
 
 
 Jinja2Template.render = instrumented_render
+
+
+@pytest.fixture(scope="session")
+def django_db_setup(django_db_setup, django_db_blocker):
+    """Seed the test database with some test data.
```
Comment (oskusalerma) [inline web/tests/conftest.py:29]: add link to the page you sent me explaining what's going on here

File: web/management/commands/add_test_data.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                security_question="security_question",
+                security_answer="security_answer",
+            )
+            self._assign_permission(user, "importer_access")
+            assign_perm("web.is_contact_of_importer", user, importer)
+            user.save()
+
+        return user
+
+    def create_admin_user(self, username):
```
Comment (oskusalerma) [inline web/management/commands/add_test_data.py:50]: maybe add a comment that this is "ICMS admin", not "django admin"

File: web/tests/domains/case/_import/fa_dfl/test_views.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+    ), "Failed to login"
+
+
+@pytest.fixture
+def dfl_app_pk(client, office, importer):
+    """Creates a fa-dfl application to be used in tests"""
+    url = reverse("import:create-fa-dfl")
+    post_data = {"importer": importer.pk, "importer_office": office.pk}
+
+    client.post(url, post_data)
```
Comment (oskusalerma) [inline web/tests/domains/case/_import/fa_dfl/test_views.py:33]: this shuold verify you get a redirect back, and get the pk from the url, not from digging around the db

File: web/tests/domains/case/_import/fa_dfl/test_views.py [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+
+def test_create_dfl_application(client, office, importer):
+    assert DFLApplication.objects.all().count() == 0
+
+    url = reverse("import:create-fa-dfl")
+    post_data = {"importer": importer.pk, "importer_office": office.pk}
+
+    resp = client.post(url, post_data)
+
+    assert DFLApplication.objects.all().count() == 1
```
Comment (oskusalerma) [inline web/tests/domains/case/_import/fa_dfl/test_views.py:46]: maybe test that count is count+1, not an absolute value?

File: web/tests/domains/case/_import/fa_dfl/test_views.py [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+    """Creates a fa-dfl application to be used in tests"""
+    url = reverse("import:create-fa-dfl")
+    post_data = {"importer": importer.pk, "importer_office": office.pk}
+
+    client.post(url, post_data)
+
+    return DFLApplication.objects.first().pk
+
+
+def test_create_dfl_application(client, office, importer):
```
Comment (oskusalerma) [inline web/tests/domains/case/_import/fa_dfl/test_views.py:38]: delete this, do all the tests in the fixture

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -58,13 +58,21 @@ Start everything using docker-compose: `make debug`
...
 
-Above script will start a PostgreSQL database and ICMS app in debug mode. In
-order to run with no debug and Gunicorn server for a production-like environment
-use: `make run`
+Above script will start a PostgreSQL database and ICMS app in debug mode.
 
 Make sure to rebuild the Docker images if new dependencies are added to the
 requirements files: `make build`.
 
+## testing
```
Comment (oskusalerma) [inline README.md:66]: capitalize

File: web/management/commands/add_test_data.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    help = """Add test data, called when running unit tests."""
+
+    def handle(self, *args, **options):
+        if settings.APP_ENV != "test":
+            raise CommandError("Can only add test data in 'test' environment!")
+
+        # Assume data is fine if the first user is already created.
+        if User.objects.filter(username="test_import_user").exists():
+            self.stdout.write("Test data already created.")
+            return
```
Comment (oskusalerma) [inline web/management/commands/add_test_data.py:23]: empty line before control flow statements

File: README.md [hunk 14]
```diff
@@ -58,13 +58,21 @@ Start everything using docker-compose: `make debug`
...
+
+To run the unit tests:
+
+`./run-tests.sh`
+
+To run tests for a single file in a non-distributed manner (faster):
+
+`./run-tests.sh --dist=no web/tests/foo/bar.by`
```
Comment (kevinetienne) [inline README.md:74]: Can you add `./run-tests.sh --create-db`? It might be useful later for newcomers.

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/tests/file_upload_handler.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
+from django.core.files.uploadhandler import TemporaryFileUploadHandler
+
+
+class DummyFileUploadHandler(TemporaryFileUploadHandler):
+    def receive_data_chunk(self, raw_data, start):
+        return super().receive_data_chunk(raw_data, start)
```
Comment (kevinetienne) [inline web/tests/file_upload_handler.py:6]: Why are you overriding this method?

File: README.md (hunk 14 above)
Comment (kevinetienne) [reply README.md:74]: I see it's already quoted in the file. Maybe you can state that more options is presented in `run-tests.sh`. Feel free to ignore this comment.

File: web/tests/file_upload_handler.py (hunk 15 above)
Comment (MattHolmes123) [reply web/tests/file_upload_handler.py:6]: Although I am doing nothing, I thought it better to be explicit about the required methods for custom file handlers.  

See here: 
https://docs.djangoproject.com/en/3.2/ref/files/uploads/#required-methods

**Review by kevinetienne — APPROVED** [review]

**Review by oskusalerma — APPROVED** [review]

**Pull request merged**